### PR TITLE
Mini-app for hackathon

### PR DESCRIPTION
**Pull request by tskisner**

### CODE REVIEW

**Comment by tskisner** [comment]
Ok, this includes a new script, "`toast_mini`" which runs a satellite simulation with pure noise and then a mapmaking.  By default it runs 10 PCG iterations and tries to fill the RAM of the job.  The data volume can be reduced with the usual `--node_mem_gb` option.  For example, running on my laptop:
```
OMP_NUM_THREADS=1 TOAST_FUNCTIME=1 mpirun -np 4 toast_mini --node_mem_gb 15.0
```
Produces:
```
TOAST INFO: To print python startup time, in the calling environment do:
            TOAST_PYTHON_START=$(date '+%Y%m%d-%H%M%S') <script> ...
TOAST INFO: TOAST version = 2.3.12.dev262
TOAST INFO: Using a maximum of 1 threads per process
TOAST INFO: Running with 4 processes at 2021-11-29 14:56:08.982728
TOAST INFO: Minimum detected per-node memory available is 52.71 GB
TOAST INFO: Setting per-node available memory to 15.00 GB as requested
TOAST INFO: Job has 1 total nodes
TOAST INFO: Threading snapshot:  Overall max = 1, Current max = 1
Memory usage system snapshot
       total :   62.555 GB  <   62.555 +-    0.000 GB  <   62.555 GB
   available :   52.714 GB  <   52.714 +-    0.000 GB  <   52.714 GB
     percent :   15.700 %   <   15.700 +-    0.000 %   <   15.700 % 
        used :    8.871 GB  <    8.871 +-    0.000 GB  <    8.871 GB
        free :   19.255 GB  <   19.255 +-    0.000 GB  <   19.255 GB
      active :    4.819 GB  <    4.819 +-    0.000 GB  <    4.819 GB
    inactive :   35.106 GB  <   35.106 +-    0.000 GB  <   35.106 GB
     buffers :    1.207 GB  <    1.207 +-    0.000 GB  <    1.207 GB
      cached :   33.221 GB  <   33.221 +-    0.000 GB  <   33.221 GB
      shared :  288.293 MB  <  288.293 +-    0.000 MB  <  288.293 MB
        slab :    2.642 GB  <    2.642 +-    0.000 GB  <    2.642 GB

TOAST WARNING: Mission start time '2021-11-29 14:56:09.037230' is not timezone-aware.  Assuming UTC.
TOAST INFO: Using automatic workflow size selection (case='auto') with 3.10 GB reserved for per process overhead.
TOAST INFO: Distribution using:
  208 detectors and 1 observations
  74879792 total samples
  1 groups of 1 nodes with 4 processes each
  14.99 GB predicted memory use   (15.00 GB available)
  ('auto' workflow size)
TOAST INFO: Focalplane has 254 detectors that span 11.367 degrees and are sampled at 100.0 Hz.
TOAST INFO: Focalplane has 208 detectors that span 9.323 degrees and are sampled at 100.0 Hz.
TOAST INFO: Using 104 hexagon-packed pixels.
TOAST INFO: Simulated telescope pointing in 0.32 s
TOAST INFO: Created default noise model in 0.00 s
TOAST INFO: Simulated detector noise in 5.23 s
TOAST INFO: MapMaker begin building flags for solver
TOAST INFO: MapMaker  finished flag building in 0.05 s
TOAST INFO: MapMaker begin build of solver covariance
TOAST INFO: MapMaker  finished build of solver covariance in 15.30 s
TOAST INFO: MapMaker Solver flags cut 0 / 74880000 = 0.00% of samples
TOAST INFO: MapMaker begin RHS calculation
TOAST INFO: MapMaker  finished RHS calculation in 12.56 s
TOAST INFO: MapMaker begin PCG solver
TOAST INFO: MapMaker initial residual = 2.736626605981918e+18, 12.12 s
TOAST INFO: MapMaker iteration    0, relative residual = 2.617093e-01, 12.04 s
TOAST INFO: MapMaker iteration    1, relative residual = 8.716097e-02, 12.19 s
TOAST INFO: MapMaker iteration    2, relative residual = 2.132956e-02, 12.14 s
TOAST INFO: MapMaker iteration    3, relative residual = 9.098786e-03, 12.51 s
TOAST INFO: MapMaker iteration    4, relative residual = 2.378607e-03, 12.47 s
TOAST INFO: MapMaker iteration    5, relative residual = 8.524663e-04, 12.29 s
TOAST INFO: MapMaker iteration    6, relative residual = 3.033495e-04, 12.10 s
TOAST INFO: MapMaker iteration    7, relative residual = 9.271196e-05, 12.45 s
TOAST INFO: MapMaker iteration    8, relative residual = 3.009590e-05, 12.33 s
TOAST INFO: MapMaker iteration    9, relative residual = 8.725570e-06, 12.37 s
TOAST INFO: MapMaker  finished solver in 135.02 s
TOAST INFO: MapMaker begin build of final binning covariance
TOAST INFO: MapMaker  finished build of final covariance in 10.37 s
TOAST INFO: MapMaker begin final map binning
TOAST INFO: MapMaker  finished final binning in 6.73 s
TOAST INFO: Wrote out_toast_mini/mapmaker_hits.fits in 1.13 s
TOAST INFO: Wrote out_toast_mini/mapmaker_map.fits in 3.92 s
TOAST INFO: MapMaker  finished output write in 5.18 s
TOAST INFO: Finished map-making in 185.20 s
```
Also in this work is some improvements to the timing plots.  Running on the outputs of the above job:
```
toast_timing_plot --file out_toast_mini/timing.csv --out out_toast_mini/timing.pdf
```
Gives the following
timing

**Comment by nestordemeure** [comment]
Seems good. Perlmutter is down right now so I couldn't try to run a rebase of my branches on top of it.

File: src/toast/_libtoast/sys.cpp [hunk 1]
```diff
@@ -458,4 +462,30 @@ void init_sys(py::module & m) {
         )");
 
     auto env = toast::Environment::get();
+
+    m.def("threading_state",
```
Comment (keskitalo) [inline src/toast/_libtoast/sys.cpp:466]: Cool!

File: src/toast/ops/noise_weight.py [hunk 2]
```diff
@@ -45,6 +45,7 @@ class NoiseWeight(Operator):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
+    @function_timer
```
Comment (keskitalo) [inline src/toast/ops/noise_weight.py:48]: Really happy to see all these timers added.

**Review by keskitalo — APPROVED** [review]
Looks good!

**Review by nestordemeure — APPROVED** [review]
Everything seems good!

File: src/libtoast/CMakeLists.txt [hunk 3]
```diff
@@ -63,21 +63,15 @@ target_include_directories(toast BEFORE PUBLIC
     $<INSTALL_INTERFACE:include>
 )
 
-# This compiler definition makes it possible to query the serial library
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/libtoast/CMakeLists.txt:66]: This was just residual code that is not used anywhere.  All MPI functionality is now in python, outside of the compiled extension.

**Pull request merged**